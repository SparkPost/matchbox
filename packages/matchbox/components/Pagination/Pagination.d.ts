export default Pagination;
declare function Pagination(props: any): any;
declare namespace Pagination {
    const displayName: string;
    const propTypes: any;
    namespace defaultProps {
        const currentPage: number;
    }
}
