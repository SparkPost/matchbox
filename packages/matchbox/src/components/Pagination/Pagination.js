import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Button, buttonsFrom } from '../Button';
import { Pager } from '../Pager';
import { MoreHoriz } from '@sparkpost/matchbox-icons';
import styles from './Pagination.module.scss';

class Pagination extends Component {
  static displayName = 'Pagination';

  static propTypes = {
    /**
     * Sets the current page number
     */
    currentPage: PropTypes.number,

    /**
     * The total number of pages
     */
    pages: PropTypes.number.isRequired,

    /**
     * The number of page buttons to display. Odd numbers look better.
     */
    pageRange: PropTypes.number.isRequired,

    /**
     * Callback when page is changed. Index passed as argument.
     */
    onChange: PropTypes.func,

    /**
     * Hides first and last buttons
     */
    marginsHidden: PropTypes.bool,

    /**
     * Flat style buttons
     */
    flat: PropTypes.bool,

    /**
     * Selected page color
     */
    selectedColor: PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red']),
  };

  static defaultProps = {
    currentPage: 1,
    selectedColor: 'orange',
  };

  state = {
    index: 0,
    hasPrevious: false,
    hasNext: true,
  };

  componentDidMount() {
    this.handlePageChange(this.props.currentPage - 1);
  }

  componentDidUpdate(prevProps) {
    const { pages, pageRange, currentPage } = this.props;
    if (
      pages !== prevProps.pages ||
      pageRange !== prevProps.pageRange ||
      currentPage !== prevProps.currentPage
    ) {
      this.handlePageChange(currentPage - 1);
    }
  }

  handlePageChange(index) {
    this.setState({
      index,
      hasPrevious: index > 0,
      hasNext: index < this.props.pages - 1,
    });

    this.props.onChange && this.props.onChange(index);
  }

  handleNext = () => {
    this.handlePageChange(this.state.index + 1);
  };

  handlePrevious = () => {
    this.handlePageChange(this.state.index - 1);
  };

  _createButtons(start, end) {
    const buttons = [];

    for (let index = start; index < end; index++) {
      buttons.push({
        content: `${index + 1}`,
        onClick: this.handlePageChange.bind(this, index),
        flat: this.props.flat,
        className: classnames(
          styles.Page,
          index === this.state.index && styles.Selected,
          index === this.state.index && styles[`color-${this.props.selectedColor}`],
        ),
      });
    }

    return buttonsFrom(buttons);
  }

  _getStart() {
    const { pages, pageRange } = this.props;
    const { index } = this.state;

    let start = 0;
    const half = Math.floor(pageRange / 2);

    if (index > half) {
      start = index - half;
    }

    if (index + half + 1 > pages) {
      start = pages - pageRange;
    }

    return start;
  }

  render() {
    const { pages, pageRange, marginsHidden, flat, className } = this.props;

    const { hasPrevious, hasNext } = this.state;

    const start = this._getStart();

    const buttonMarkup = () => {
      if (pages <= pageRange) {
        return this._createButtons(0, pages);
      } else {
        return this._createButtons(start, start + pageRange);
      }
    };

    const firstButton =
      !marginsHidden && start > 1 ? (
        <span>
          <Button flat={flat} className={styles.Start} onClick={() => this.handlePageChange(0)}>
            1
          </Button>
          <MoreHoriz className={styles.Ellipse} style={{ marginTop: '-0.2em' }} />
        </span>
      ) : null;

    const lastButton =
      !marginsHidden && start + pageRange < pages ? (
        <span>
          <MoreHoriz className={styles.Ellipse} style={{ marginTop: '-0.2em' }} />
          <Button
            flat={flat}
            className={styles.End}
            onClick={() => this.handlePageChange(pages - 1)}
          >
            {pages}
          </Button>
        </span>
      ) : null;

    return (
      <div className={classnames(styles.Pagination, className)}>
        <Pager.Previous flat={flat} onClick={this.handlePrevious} disabled={!hasPrevious} />
        {firstButton}
        <span className={styles.Pages}>{buttonMarkup()}</span>
        {lastButton}
        <Pager.Next flat={flat} onClick={this.handleNext} disabled={!hasNext} />
      </div>
    );
  }
}

export default Pagination;
