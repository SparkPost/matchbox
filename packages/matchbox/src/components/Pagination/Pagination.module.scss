@import '../../styles/config';

.Pagination {
  margin-bottom: spacing();
}

.Back {
  margin-right: spacing(small);
}

.Next {
  margin-left: spacing(small);
}

.Pages {
  display: inline-block;
}

.Page, .Start, .End {
  color: color(gray, 3);
  min-width: rem(42);
}

.Selected {
  font-weight: 600;
  pointer-events: none;
}

.color-orange {
  color: color(orange);
}

.color-blue {
  color: color(blue, dark);
}

.color-navy {
  color: color(navy, light);
}

.color-purple {
  color: color(purple, light);
}

.color-red {
  color: color(red, light);
}

.Ellipse {
  margin-left: spacing(small);
  margin-right: spacing(small);
  fill: color(gray, 6);
}
