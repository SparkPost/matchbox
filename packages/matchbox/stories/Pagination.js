import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Pagination } from '../src';

export default storiesOf('Pagination', module)
  .addDecorator((getStory) => (
    <StoryContainer>{ getStory() }</StoryContainer>
  ))

  .add('with initial index', withInfo()(() => (
    <Pagination
      pages={4}
      pageRange={4}
      initialIndex={3}
      onChange={action('Page Changed')}
    />
  )))

  .add('with lots of pages', withInfo()(() => (
    <Pagination
      pages={30}
      pageRange={7}
      selectedColor='navy'
      onChange={action('Page Changed')}
    />
  )));
