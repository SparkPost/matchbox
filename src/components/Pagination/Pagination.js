import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Button, buttonsFrom } from '../Button';
import { Icon } from '../Icon';
import styles from './Pagination.module.scss';

class Pagination extends Component {
  static propTypes = {
    /**
     * The total number of pages
     */
    pages: PropTypes.number.isRequired,
    /**
     * The number of page buttons to display. Odd numbers look better.
     */
    pageRange: PropTypes.number.isRequired,
    /**
     * Callback when page is changed. Index passed as argument.
     */
    onChange: PropTypes.func,
    /**
     * Sets page index on component mount.
     */
    initialIndex: PropTypes.number.isRequired
  };

  static defaultProps = {
    initialIndex: 0
  };

  constructor(props) {
    super(props);
    this.state = {
      index: 0,
      hasPrevious: false,
      hasNext: true
    };
  }

  componentWillMount() {
    this.handlePageChange(this.props.initialIndex);
  }

  handlePageChange(index) {
    this.setState({
      index,
      hasPrevious: index > 0,
      hasNext: index < this.props.pages - 1
    });

    this.props.onChange && this.props.onChange(index);
  }

  handleNext() {
    this.handlePageChange(this.state.index + 1);
  }

  handlePrevious() {
    this.handlePageChange(this.state.index - 1);
  }

  _createButtons(start, end) {
    let buttons = [];

    for (let index = start; index < end; index++) {
      buttons.push({
        content: `${index + 1}`,
        onClick: this.handlePageChange.bind(this, index),
        className: classnames(styles.Page, index === this.state.index && styles.Selected)
      });
    }

    return buttonsFrom(buttons);
  }

  render() {
    const {
      pages,
      pageRange
    } = this.props;

    const {
      hasPrevious,
      hasNext
    } = this.state;

    const buttonMarkup = () => {
      if (pages <= pageRange) {
        return this._createButtons(0, pages);
      } else {
        const half = Math.floor(pageRange / 2);
        let start = 0;

        if (this.state.index > half) {
          start = this.state.index - half;
        }

        if (this.state.index + half + 1 > pages) {
          start = pages - pageRange;
        }

        return this._createButtons(start, start + pageRange);
      }
    };

    return (
      <div className={styles.Pagination}>
        <Button
          className={styles.Back}
          onClick={() => this.handlePrevious()}
          disabled={!hasPrevious} >
          <Icon name='ArrowBack' size={16} />
        </Button>
        <span className={styles.Pages}>{ buttonMarkup() }</span>
        <Button
          className={styles.Next}
          onClick={() => this.handleNext()}
          disabled={!hasNext} >
          <Icon name='ArrowForward' size={16} />
        </Button>
      </div>
    );
  }
};

export default Pagination;
