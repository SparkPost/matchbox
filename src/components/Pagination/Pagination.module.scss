@import '../../styles/config';

.Pagination {
  margin-bottom: spacing();
}

.Back {
  margin-right: spacing(small);
}

.Next {
  margin-left: spacing(small);
}

.Pages {
  display: inline-block;
}

.Page, .Start, .End {
  width: rem(39);
}

.Selected {
  font-weight: 700;
  color: color(orange);
}

.Ellipse {
  margin-left: spacing(small);
  margin-right: spacing(small);
  fill: color(gray, 6);
}
