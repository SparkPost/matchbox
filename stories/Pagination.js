import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Pagination } from '../src';

export default storiesOf('Pagination', module)
  .add('Default', () => (
    <StoryContainer>
      <Pagination
        pages={4}
        pageRange={4}
        onChange={action('Page Changed')}
      />

      <Pagination
        pages={13}
        pageRange={7}
        initialIndex={4}
        onChange={action('Page Changed')}
      />
    </StoryContainer>
  ));
