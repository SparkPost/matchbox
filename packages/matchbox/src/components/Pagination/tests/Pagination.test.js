import React from 'react';
import { shallow } from 'enzyme';

import Pagination from '../Pagination';

describe('Pagination', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      currentPage: 1,
      pages: 10,
      pageRange: 3,
      selectedColor: 'blue'
    };

    wrapper = shallow(<Pagination {...props}/>);
  });

  it('renders pagination', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('invokes onChange on page', () => {
    const fn = jest.fn();
    wrapper.setProps({ onChange: fn });
    wrapper.find('Button').at(0).simulate('click');
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('renders pagination with marginsHidden true', () => {
    wrapper.setProps({ marginsHidden: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders pagination with flat buttons', () => {
    wrapper.setProps({ flat: true });
    expect(wrapper).toMatchSnapshot();
  });
});
