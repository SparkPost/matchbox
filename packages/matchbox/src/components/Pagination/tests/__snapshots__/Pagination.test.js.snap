// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders pagination 1`] = `
<div
  className="Pagination"
>
  <Button
    className="Back"
    disabled={true}
    onClick={[Function]}
    size="default"
  >
    <ArrowBack
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
  <span
    className="Pages"
  >
    <Button.Group>
      <Button
        className="Page Selected color-blue"
        key="0"
        onClick={[Function]}
        size="default"
      >
        1
      </Button>
      <Button
        className="Page"
        key="1"
        onClick={[Function]}
        size="default"
      >
        2
      </Button>
      <Button
        className="Page"
        key="2"
        onClick={[Function]}
        size="default"
      >
        3
      </Button>
    </Button.Group>
  </span>
  <span>
    <MoreHoriz
      className="Ellipse"
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
    <Button
      className="End"
      onClick={[Function]}
      size="default"
    >
      10
    </Button>
  </span>
  <Button
    className="Next"
    disabled={false}
    onClick={[Function]}
    size="default"
  >
    <ArrowForward
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
</div>
`;

exports[`Pagination renders pagination with flat buttons 1`] = `
<div
  className="Pagination"
>
  <Button
    className="Back"
    disabled={true}
    flat={true}
    onClick={[Function]}
    size="default"
  >
    <ArrowBack
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
  <span
    className="Pages"
  >
    <Button.Group>
      <Button
        className="Page Selected color-blue"
        flat={true}
        key="0"
        onClick={[Function]}
        size="default"
      >
        1
      </Button>
      <Button
        className="Page"
        flat={true}
        key="1"
        onClick={[Function]}
        size="default"
      >
        2
      </Button>
      <Button
        className="Page"
        flat={true}
        key="2"
        onClick={[Function]}
        size="default"
      >
        3
      </Button>
    </Button.Group>
  </span>
  <span>
    <MoreHoriz
      className="Ellipse"
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
    <Button
      className="End"
      flat={true}
      onClick={[Function]}
      size="default"
    >
      10
    </Button>
  </span>
  <Button
    className="Next"
    disabled={false}
    flat={true}
    onClick={[Function]}
    size="default"
  >
    <ArrowForward
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
</div>
`;

exports[`Pagination renders pagination with marginsHidden true 1`] = `
<div
  className="Pagination"
>
  <Button
    className="Back"
    disabled={true}
    onClick={[Function]}
    size="default"
  >
    <ArrowBack
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
  <span
    className="Pages"
  >
    <Button.Group>
      <Button
        className="Page Selected color-blue"
        key="0"
        onClick={[Function]}
        size="default"
      >
        1
      </Button>
      <Button
        className="Page"
        key="1"
        onClick={[Function]}
        size="default"
      >
        2
      </Button>
      <Button
        className="Page"
        key="2"
        onClick={[Function]}
        size="default"
      >
        3
      </Button>
    </Button.Group>
  </span>
  <Button
    className="Next"
    disabled={false}
    onClick={[Function]}
    size="default"
  >
    <ArrowForward
      size={16}
      style={
        Object {
          "marginTop": "-0.2em",
        }
      }
    />
  </Button>
</div>
`;
